### PR TITLE
feature: Add BracePairFoldingRegionProvider and tests (#3795)

### PR DESCRIPTION
**Pull request by ArneDeutsch**
Implemented an alternative folding provider that works on the base of IBracePairProvider. It does NOT fold the EObjects but instead folds the regions based on the brace pairs (default are the pairs {}, () and [] ). Users have to use BracePairFoldingRegionProvider instead of DefaultFoldingRegionProvider and can adapt behaviour by using different implementation of IBracePairProvider.

### CODE REVIEW

**Comment by ArneDeutsch** [comment]
The failure looks like an unstable test to me (unrelated):

Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.336 s <<< FAILURE! -- in org.eclipse.xtext.ui.tests.editor.hover.AbstractProblemHoverTest
org.eclipse.xtext.ui.tests.editor.hover.AbstractProblemHoverTest.testAnnotations -- Time elapsed: 0.192 s <<< FAILURE!
java.lang.AssertionError: expected:<1> but was:<0>

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/folding/BracePairFoldingRegionProvider.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import org.eclipse.xtext.util.ITextRegion;
+
+import com.google.inject.Inject;
+
+/**
+ * Calculate the folding regions from the brace pairs and the comments. The default regions from the EObjects are skipped.
+ * 
+ * @author Arne Deutsch - Initial contribution and API
+ */
+public class BracePairFoldingRegionProvider extends DefaultFoldingRegionProvider {
```
Comment (tamasmiklossy) [inline org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/folding/BracePairFoldingRegionProvider.java:32]: missing `@since`

Comment (ArneDeutsch) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/folding/BracePairFoldingRegionProvider.java:32]: Added

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/folding/BracePairFoldingRegionProvider.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		}
+		ICompositeNode root = parseResult.getRootNode();
+		if (root == null) {
+			return;
+		}
+		computeBracePairFolding(root, acceptor);
+	}
+
+	/**
+	 * The folding regions are calculated by itrerating the tokens once and checking for matching opening/closing brace pairs. The
```
Comment (tamasmiklossy) [inline org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/folding/BracePairFoldingRegionProvider.java]: Typo: `itrerating` -> `iterating`
```suggestion
	 * The folding regions are calculated by iterating the tokens once and checking for matching opening/closing brace pairs. The
```

File: org.eclipse.xtext.ui.tests/tests/org/eclipse/xtext/ui/tests/editor/folding/BracePairFoldingRegionProviderTest.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	@Test
+	public void brokenInputFoldsBestEffort() throws Exception {
+		testFoldingRegions("""
+				root {
+				[>  child {
+				  }<]
+				""");
+	}
+
+}
```
Comment (tamasmiklossy) [inline org.eclipse.xtext.ui.tests/tests/org/eclipse/xtext/ui/tests/editor/folding/BracePairFoldingRegionProviderTest.java:204]: @ArneDeutsch Can you please add the following test cases to verify if the folding works properly if there is only hidden tokens (empty lines or comments) between the curly braces:

```
[>root() {
}<]
```

```
[>root() {

}<]
```

```
[>root() {


}<]
```

```
[>root() {



}<]
```

```
[>root() {




}<]
```

```
[>root() {
[>	/*
	 * multi-line comment
	 */<]
}<]
```

```
[>root() {
[>	/**
	 * javadoc comment
	 */<]
}<]
```

```
[>root() {
	// single line comment1

	// single line comment2
}<]
```

```
[>root() {
	// single line comment

[>	/*
	 * multi-line comment
	 */<]
}<]
```

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/folding/BracePairFoldingRegionProvider.java (hunk 2 above)
Comment (ArneDeutsch) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/folding/BracePairFoldingRegionProvider.java]: Corrected

File: org.eclipse.xtext.ui.tests/tests/org/eclipse/xtext/ui/tests/editor/folding/BracePairFoldingRegionProviderTest.java (hunk 3 above)
Comment (ArneDeutsch) [reply org.eclipse.xtext.ui.tests/tests/org/eclipse/xtext/ui/tests/editor/folding/BracePairFoldingRegionProviderTest.java:204]: Added

**Comment by tamasmiklossy** [comment]
Thank you @ArneDeutsch : LGTM, approved from my side :-)
